Model API response and add reducer (#185)
pupilfirst/wd301-scripts: handling-advanced-ui-events/drag-and-drop/response-types-and-reducer/README.md

# Text

In this lesson, we will create TypeScript types to model the API response.

Let's take a look at the API response for listing a task.

> Visit [List task API](https://wd301-api.pupilfirst.school/#/Tasks/get_projects__projectId__tasks)

> We can see the reponse is of the following shape.

Reviewer: ```suggestion
We can see the response is of the following shape.
```

```json
{

>   "coloumns": {

Reviewer: ```suggestion
  "columns": {
```

    "pending": {
      "id": "pending",
      "title": "Pending",
      "taskIDs": ["2"]
    },
    "in_progress": {
      "id": "in_progress",
      "title": "In progress",
      "taskIDs": ["1"]
    },
    "done": {
      "id": "done",
      "title": "Done",
      "taskIDs": []
    }
  },
  "tasks": {
    "1": {
      "id": 1,
      "title": "Sample Task",
      "description": "Sample description about the task which is to be completed",
      "dueDate": "",
      "state": "in_progress",
      "assignee": null,
      "assignedUserName": null
    },
    "2": {
      "id": 2,
      "title": "Another Sample Task",
      "description": "Sample description about the task which is to be completed",
      "dueDate": "",
      "state": "pending",
      "assignee": null,
      "assignedUserName": null
    }
  },

>   "coloumnOrder": ["pending", "in_progress", "done"]

Reviewer: ```suggestion
  "columnOrder": ["pending", "in_progress", "done"]
```

}
```

> We have three coloumns ie, `pending`, `in_progress`, and `done`. We have `coloumnOrder`, which can be used to control in what order the lists must be rendered. Then we have the tasks which can be accessed using their `id`.

Reviewer: ```suggestion
We have three columns ie, `pending`, `in_progress`, and `done`. We have `columnOrder`, which can be used to control in what order the lists must be rendered. Then we have the tasks which can be accessed using their `id`.
```

Now, let's create types to model this shape.

We will use a `union` type to model the coloumns.

Open `src/context/task/types.ts` file in VS Code and add the following entry to it.

```ts

> export type AvailableColoumns = "pending" | "in_progress" | "done";

Reviewer: ```suggestion
export type AvailableColumns = "pending" | "in_progress" | "done";
```

```

> Each entry in a coloumn ie,

Reviewer: ```suggestion
Each entry in a column ie,
```

```json
{
  "id": "pending",
  "title": "Pending",
  "taskIDs": ["2"]
}
```

can be modelled as follows.

```ts

> export type ColoumnData = {

Reviewer: ```suggestion
export type ColumnData = {
```

  id: string;
  title: string;
  taskIDs: string[];
};
```

> Then the whole `coloumn` key in the response can be modelled as:

Reviewer: ```suggestion
Then the whole `column` key in the response can be modelled as:
```
